spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/amqp/RabbitAutoConfiguration.java

```diff
@@ -93,73 +92,48 @@
 public class RabbitAutoConfiguration {
 
 	@Configuration(proxyBeanMethods = false)
-	@ConditionalOnMissingBean(ConnectionFactory.class)
 	protected static class RabbitConnectionFactoryCreator {
 
 		@Bean
-		public CachingConnectionFactory rabbitConnectionFactory(RabbitProperties properties,
+		@ConditionalOnMissingBean
+		RabbitConnectionFactoryBeanConfigurer rabbitConnectionFactoryBeanConfigurer(RabbitProperties properties,
 				ResourceLoader resourceLoader, ObjectProvider<CredentialsProvider> credentialsProvider,
-				ObjectProvider<CredentialsRefreshService> credentialsRefreshService,
-				ObjectProvider<ConnectionNameStrategy> connectionNameStrategy,
+				ObjectProvider<CredentialsRefreshService> credentialsRefreshService) {
+			RabbitConnectionFactoryBeanConfigurer configurer = new RabbitConnectionFactoryBeanConfigurer();
+			configurer.setRabbitProperties(properties);
+			configurer.setResourceLoader(resourceLoader);
+			configurer.setCredentialsProvider(credentialsProvider.getIfUnique());
+			configurer.setCredentialsRefreshService(credentialsRefreshService.getIfUnique());
+			return configurer;
+		}
+
+		@Bean
+		@ConditionalOnMissingBean
+		CachingConnectionFactoryConfigurer rabbitConnectionFactoryConfigurer(RabbitProperties rabbitProperties,
+				ObjectProvider<ConnectionNameStrategy> connectionNameStrategy) {
+			CachingConnectionFactoryConfigurer configurer = new CachingConnectionFactoryConfigurer();
+			configurer.setRabbitProperties(rabbitProperties);
+			configurer.setConnectionNameStrategy(connectionNameStrategy.getIfUnique());
+			return configurer;
+		}
+
+		@Bean
+		@ConditionalOnMissingBean(ConnectionFactory.class)
+		CachingConnectionFactory rabbitConnectionFactory(
```

Comment:
This is a breaking change:
1. visibility public->package (talking about that in https://github.com/spring-projects/spring-boot/pull/26982/files#r655007725)
2. method signature change (params)

We could probably leave the old method and deprecate and then do this additively. I just wanted to put this out there before investing in it though.

Reply:
A breaking change here is fine as we don't consider the `@Bean` methods on an auto-configuration class to be public API.